feat: multi interface support

therishidesai opened the pull request:
Add the ability to support joining the multicast group on multiple interfaces. In some environments, there are many network interfaces that can support multicast discovery and so we should allow `swarm_discovery` users to use all of the available interfaces.

matheus23 commented on src/lib.rs:
```diff
@@ -356,13 +360,26 @@ impl Discoverer {
     /// Changing the configuration is done by stopping the discovery and starting a new one.
     pub fn spawn(self, handle: &Handle) -> Result<DropGuard, SpawnError> {
         let _entered = handle.enter();
-        let sockets = Sockets::new(self.class)?;
```
Maybe it would be better if you could pass in the sockets that swarm discovery is supposed to use from the outside, and then all this interface enumeration & selection can live outside.
This might result in a much smaller PR that's way easier to review & accept.

therishidesai commented:
> Given that there's no `flake.nix` nor `flake.lock` files in this repo yet, I suspect the owner doesn't use nix, so please try moving your nix configuration elsewhere so it doesn't end up in this PR ✌️

It isn't my personal nix configuration. Just contributing a development shell for others to use. If @rkuhn wants me to remove this then I can but this is a common thing many projects will have to give people a reproducible dev environment

therishidesai commented:
Converting to a draft since this doesn't quite work correctly and also looking into @matheus23 suggestion to move the interface detection out of this library

therishidesai commented:
This is ready for review now. Tested it with various different connection and it is working correctly. 

@matheus23 I didn't make it so the user passes in Sockets but instead they can add a list of multicast interface IPs that we then bind to. This achieves the goal of removing the interface logic from the library while also keeping backwards compatibility for the API.

therishidesai commented:
> Thanks for contributing, this looks useful! Would it make sense to figure out IPv6 as well or is that for another day/PR? I’m currently in a different context and have no good handle on how much effort it would be beyond figuring out a sensible API. Perhaps have different builder methods for the two address families, one accepting IPv4 addresses and one accepting interface indices?

Yeah,I was thinking of refactoring this to just be IPv4 only. I also have a use case where I would want to dynamically add interfaces to connect. Should have something up later today.

therishidesai commented:
@rkuhn So I cleaned up the API to only be supported on IPV4 and just no-op on IPV6. I also added the ability to dynamically add interfaces or remove interfaces. I specifically have a use case where an interface may appear in the middle of my program and I don't want to restart the Discoverer.

therishidesai commented:
@rkuhn I was wondering if you had a chance to take another look at this. I would like to get this PR in so I can also update iroh to use it and not have to use a fork

therishidesai commented on src/socket.rs:
```diff
@@ -204,15 +233,36 @@ pub fn socket_v6() -> Result<UdpSocket, SocketError> {
 pub struct Sockets {
     v4: Option<Arc<UdpSocket>>,
     v6: Option<Arc<UdpSocket>>,
+    interface_sockets_v4: Arc<RwLock<Vec<(IpAddr, Arc<UdpSocket>)>>>,
```
If I use a map would you be ok with bringing in the rust [dashmap](https://github.com/xacrimon/dashmap) crate? This is a concurrent hashmap implementation which removes the need for wrapping in an `Arc<RwLock>`

rkuhn commented on src/socket.rs:
```diff
@@ -204,15 +221,34 @@ pub fn socket_v6() -> Result<UdpSocket, SocketError> {
+    pub fn new(class: IpClass, multicast_interfaces: Vec<Ipv4Addr>) -> Result<Self, SocketError> {
+        // Create interface-specific sockets for multi-interface mode
+        let mut interface_sockets_v4 = HashMap::new();
+        if !multicast_interfaces.is_empty() {
```
This `if` is extraneous and could be removed.

rkuhn commented on src/socket.rs:
```diff
@@ -240,6 +277,51 @@ impl Sockets {
+        let mut interfaces = self.interface_sockets_v4.write().unwrap();
+
+        if interfaces.contains_key(&addr) {
+            interfaces.remove(&addr);
```
same here as above, but the other way around: dropping the socket could block for some time, right? so better release the lock before dropping the socket

rkuhn commented on src/socket.rs:
```diff
@@ -240,6 +277,51 @@ impl Sockets {
+    /// Get all interface addresses that have sockets
+    pub fn get_all_interface_addresses_v4(&self) -> Vec<Ipv4Addr> {
+        let interfaces = self.interface_sockets_v4.read().unwrap();
+        interfaces.iter().map(|(addr, _)| *addr).collect()
```
```suggestion
        interfaces.keys().copied().collect()
```

rkuhn reviewed (commented):
Yes, basically looking good, with only nitpicks left. Sorry for the delay, again. Do you want me to apply the final touches?
